### PR TITLE
Fix Issue #1 $count is evaluated prematurely at the call queryOptions.ApplyTo

### PR DESCRIPTION
**Pull request by VikingsFan**

### CODE REVIEW

File: OData/src/System.Web.OData/Extensions/HttpRequestMessageProperties.cs [hunk 1]
```diff
@@ -146,6 +165,12 @@ public Routing.ODataPath Path
                     return (long)totalCount;
                 }
 
+                if (this.TotalCountFunc != null)
+                {
+                    _request.Properties[TotalCountKey] = this.TotalCountFunc();
```
Comment (lewischeng-ms) [inline OData/src/System.Web.OData/Extensions/HttpRequestMessageProperties.cs]: Cache the result of `_request.Properties[TotalCountKey]` to a local variable to increase performance.

File: OData/test/UnitTest/System.Web.OData.Test/OData/Query/CountQueryOptionTest.cs [hunk 2]
```diff
@@ -134,6 +134,26 @@ public void GetEntityCount_ReturnsNull_IfValueIsFalse()
         }
 
         [Fact]
+        public void GetEntityCountFunc_ReturnsFunc_IfValueIsTrue()
+        {
+            // Arrange
+            var countOption = new CountQueryOption("true", _context);
```
Comment (lewischeng-ms) [inline OData/test/UnitTest/System.Web.OData.Test/OData/Query/CountQueryOptionTest.cs]: Also test the case where $count not present?

**Comment by lewischeng-ms** [comment]
Looks good to me basically:P

File: OData/test/UnitTest/System.Web.OData.Test/OData/Query/CountQueryOptionTest.cs (hunk 2 above)
Comment (VikingsFan) [inline OData/test/UnitTest/System.Web.OData.Test/OData/Query/CountQueryOptionTest.cs]: Added Thanks!

File: OData/test/E2ETest/WebStack.QA.Test.OData/ODataCountTest/CountTest.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Get, requestUri);
+
+            // Act
+            var response = await Client.SendAsync(request);
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            string result = response.Content.ReadAsStringAsync().Result;
+            int tryParseInt;
+            bool tryResult = Int32.TryParse(result, out tryParseInt);
```
Comment (xuzhg) [inline OData/test/E2ETest/WebStack.QA.Test.OData/ODataCountTest/CountTest.cs]: why do you try to parse the total payload? As I know, you should only to test the @odata.count.

**Comment by xuzhg** [comment]
others look good to me. Thanks.

File: OData/test/E2ETest/WebStack.QA.Test.OData/ODataCountTest/CountTest.cs (hunk 3 above)
Comment (VikingsFan) [inline OData/test/E2ETest/WebStack.QA.Test.OData/ODataCountTest/CountTest.cs]: Thanks, updated

**Pull request merged**